libbeat/reader/multiline/counter.go

```diff
@@ -0,0 +1,131 @@
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package multiline
+
+import (
+	"github.com/elastic/beats/v7/libbeat/reader"
+)
+
+type counterReader struct {
+	reader    reader.Reader
+	state     func(*counterReader) (reader.Message, error)
+	msgBuffer *messageBuffer
+}
+
+func newMultilineCountReader(
+	r reader.Reader,
+	separator string,
+	maxBytes int,
+	config *Config,
+) (reader.Reader, error) {
+	return &counterReader{
+		reader:    r,
+		state:     (*counterReader).readFirst,
+		msgBuffer: newMessageBuffer(maxBytes, config.LinesCount, []byte(separator), config.SkipNewLine),
+	}, nil
+}
+
+// Next returns next multi-line event.
+func (cr *counterReader) Next() (reader.Message, error) {
+	return cr.state(cr)
+}
+
+func (cr *counterReader) readFirst() (reader.Message, error) {
+	for {
+		message, err := cr.reader.Next()
+		if err != nil {
+			return message, err
+		}
+
+		if message.Bytes == 0 {
+			continue
+		}
+
+		cr.msgBuffer.startNewMessage(message)
+		if cr.isEnoughLines() {
+			msg := cr.msgBuffer.finalize()
+			return msg, nil
+		}
+
+		cr.setState((*counterReader).readNext)
+		return cr.readNext()
+	}
+}
+
+func (cr *counterReader) readNext() (reader.Message, error) {
+	for {
+		message, err := cr.reader.Next()
+		if err != nil {
+			// handle error without any bytes returned from reader
+			if message.Bytes == 0 {
+				// no lines buffered -> return error
+				if cr.msgBuffer.isEmpty() {
+					return reader.Message{}, err
+				}
+
+				// lines buffered, return multiline and error on next read
+				msg := cr.msgBuffer.finalize()
+				cr.msgBuffer.setErr(err)
+				cr.setState((*counterReader).readFailed)
+				return msg, nil
+			}
+
+			// handle error with some content being returned by reader and
+			// line matching multiline criteria or no multiline started yet
+			if cr.msgBuffer.isEmptyMessage() {
+				cr.msgBuffer.addLine(message)
+
+				// return multiline and error on next read
+				msg := cr.msgBuffer.finalize()
+				cr.msgBuffer.setErr(err)
+				cr.setState((*counterReader).readFailed)
+				return msg, nil
+			}
+		}
+
+		// add line to current multiline event
+		cr.msgBuffer.addLine(message)
+		if cr.isEnoughLines() {
+			msg := cr.msgBuffer.finalize()
+			cr.setState((*counterReader).readFirst)
+			return msg, nil
+		}
+
+	}
+}
+
+func (cr *counterReader) isEnoughLines() bool {
+	return cr.msgBuffer.numLines == cr.msgBuffer.maxLines
```

Comment:
This makes the state machine depend on the msgBuffer, but also restricts the capabilities of the message buffer (it must be configured that that the total number of lines is >= LinesCount). Introducing our own counters in the state machine allows us to separate 'truncation' from the business logic, also allowing configuration with `MaxLines < LinesCount` (e.g. collect 10 lines, but only report first line).

If MaxLines is not set, it should indeed be set to LinesCount